cmd/multiFaAcceleration/multiFaAcceleration.go

```diff
@@ -0,0 +1,504 @@
+		}
+		Q = nextQ
+		i++
+	}
+	if i >= maxIteration {
+		log.Fatalf("Failed to converge on a tree with these distances. D01: %f, D02: %f, D03: %f, D12: %f, D13: %f, D23: %f.", d.D01, d.D02, d.D03, d.D12, d.D13, d.D23)
+	}
+	return answer
+}
+
+//a helper function of alternatingLeastSquares. Calculates the optimal branch lengths for the three branches in a subtree.
+func optimizeSubtree(sub *SubTree, s Settings) (float64, float64, float64) {
+	sub.va = (sub.Dab + sub.Dac - sub.Dbc) / 2.0
+	sub.vb = (sub.Dab + sub.Dbc - sub.Dac) / 2.0
+	sub.vc = (sub.Dac + sub.Dbc - sub.Dac) / 2.0
+
+	if s.AllowNegative {
+		return sub.va, sub.vb, sub.vc
+	}
+	if sub.va < 0 && sub.vb < 0 && sub.vc < 0 {
+		if s.Verbose {
+			log.Printf("WARNING: All branches are negative.")
+		}
+		sub.va, sub.vb, sub.vc = 0, 0, 0
+	} else if sub.va < 0 && sub.vb < 0 {
+		sub.va = 0
+		sub.vb = 0
+		sub.vc = nonNegativeApproximation(sub.Dac, sub.Dbc, sub.va, sub.vb, s)
+	} else if sub.va < 0 && sub.vc < 0 {
+		sub.va = 0
+		sub.vc = 0
+		sub.vb = nonNegativeApproximation(sub.Dbc, sub.Dab, sub.vc, sub.va, s)
+	} else if sub.vb < 0 && sub.vc < 0 {
+		sub.vb = 0
+		sub.vc = 0
+		sub.va = nonNegativeApproximation(sub.Dab, sub.Dac, sub.vb, sub.vc, s)
+	} else if sub.va < 0 {
+		sub.va = 0
+		sub.vb = nonNegativeApproximation(sub.Dab, sub.Dbc, sub.va, sub.vc, s)
+		sub.vc = nonNegativeApproximation(sub.Dac, sub.Dbc, sub.va, sub.vb, s)
+	} else if sub.vb < 0 {
+		sub.vb = 0
+		sub.va = nonNegativeApproximation(sub.Dab, sub.Dac, sub.vb, sub.vc, s)
+		sub.vc = nonNegativeApproximation(sub.Dab, sub.Dbc, sub.va, sub.vc, s)
+	} else if sub.vc < 0 {
+		sub.vc = 0
+		sub.va = nonNegativeApproximation(sub.Dab, sub.Dac, sub.vb, sub.vc, s)
+		sub.vb = nonNegativeApproximation(sub.Dab, sub.Dbc, sub.va, sub.vc, s)
+	}
+	return sub.va, sub.vb, sub.vc
+}
+
+//If we constrain branch lengths to be nonNegative, we apply this correction when the minimum Q is achieved at negative branch lengths for a subtree.
+func nonNegativeApproximation(d1 float64, d2 float64, v1 float64, v2 float64, s Settings) float64 {
+	if d1 == 0 {
+		if d2 == 0 {
+			return numbers.MaxFloat64(0, (s.ZeroDistanceWeightConstant*(d1-v1)+s.ZeroDistanceWeightConstant*(d2-v2))/(2*s.ZeroDistanceWeightConstant))
+		} else {
+			return numbers.MaxFloat64(0, s.ZeroDistanceWeightConstant*(d1-v1)+(1.0/math.Pow(d2, 2)*(d2-v2))) / (s.ZeroDistanceWeightConstant + (1.0 / math.Pow(d2, 2)))
+		}
+	} else if d2 == 0 {
+		return numbers.MaxFloat64(0, (1.0/(math.Pow(d1, 2))*(d1-v1)+s.ZeroDistanceWeightConstant*(d2-v2))/((1.0/math.Pow(d1, 2))+s.ZeroDistanceWeightConstant))
+	}
+	return numbers.MaxFloat64(0, (1.0/(math.Pow(d1, 2))*(d1-v1)+(1.0/math.Pow(d2, 2)*(d2-v2)))/((1.0/math.Pow(d1, 2))+(1.0/math.Pow(d2, 2))))
+}
+
+//Reduce the four species tree to the subtree containing species 0, 1, and the ancestor of 2/3.
+func pruneLeft(d Distances, b BranchLengths, sub *SubTree, s Settings) {
+	sub.Dab = d.D01
+	if d.D03 == 0 {
```

Comment:
Similar to my above comment, I think all these long and separate equations makes it difficult to pick out bugs.  Can we make the weights variables, which get set with if-else statements and then we just have one equation?